### PR TITLE
feat(VList): render subheaders and dividers through items prop

### PR DESCRIPTION
**Pull request by nekosaur**
## Description
add ability to render list subheader and dividers through items prop on v-list

## Markup:

<details>

```vue
<template>
  <v-app>
    <v-list v-model:active="active" v-model:opened="open" :items="items" />
    <div>
      {{ open }}
    </div>
    <div>
      {{ active }}
    </div>
  </v-app>
</template>

<script>
  export default {
    name: 'Playground',
    data: () => ({
      open: null,
      active: null,
      items: [
        {
          $type: 'subheader',
          title: 'Foo',
        },
        {
          $type: 'item',
          title: 'Item',
          value: 'item',
          subtitle: 'Subtitle',
        },
        {
          $type: 'divider',
        },
        {
          $type: 'item',
          title: 'Item #2',
          subtitle: 'Another subtitle',
          $children: [
            {
              $type: 'subheader',
              title: 'Test',
              value: 'test',
            },
            {
              title: 'Hello',
              value: 'hello',
            },
          ],
        },
      ],
      loadedItems: [
        {
          value: 'root',
          text: 'root',
          loaded: false,
          children: [],
        },
      ],
      opened: [],
      isCtrl: false,
      count: 1,
    }),
  }
</script>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by nekosaur** [comment]
my initial idea for the items structure was what is now the `InternalListItem`, but @johnleider wanted something a bit more simple/flatter. Flattening the structure means that extra user properties might clash with type and/or children, which is why they are currently prefixed with `$`. It's not a perfect solution, but should avoid most conflicts

**Comment by johnleider** [comment]
> my initial idea for the items structure was what is now the `InternalListItem`, but @johnleider wanted something a bit more simple/flatter. Flattening the structure means that extra user properties might clash with type and/or children, which is why they are currently prefixed with `$`. It's not a perfect solution, but should avoid most conflicts

It is a pretty well-accepted pattern for private internals in software. As long as we document it properly, there should be no issues.

File: packages/vuetify/src/components/VList/VListChildren.tsx [hunk 1]
```diff
@@ -24,25 +26,28 @@ export const VListChildren = genericComponent<new <T extends ListItem>() => {
...
-      const { value, ...rest } = item
+    return () => slots.default?.() ?? props.items?.map(({ children, props: itemProps, type }) => {
+      if (type === 'divider') return <VDivider {...itemProps} />
+
+      if (type === 'subheader') return <VListSubheader {...itemProps} v-slots={ slots } />
+
       return children ? (
         <VListGroup
-          value={value}
+          value={itemProps?.value}
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/VListChildren.tsx:40]:
```suggestion
          value={ itemProps?.value }
```

File: packages/vuetify/src/components/VList/VListChildren.tsx [hunk 2]
```diff
@@ -24,25 +26,28 @@ export const VListChildren = genericComponent<new <T extends ListItem>() => {
...
   props: {
-    items: Array as Prop<ListItem[]>,
+    items: Array as Prop<InternalListItem[]>,
   },
 
   setup (props, { slots }) {
-    return () => slots.default?.() ?? props.items?.map(({ children, ...item }) => {
-      const { value, ...rest } = item
+    return () => slots.default?.() ?? props.items?.map(({ children, props: itemProps, type }) => {
+      if (type === 'divider') return <VDivider {...itemProps} />
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/VListChildren.tsx:34]:
```suggestion
      if (type === 'divider') return <VDivider { ...itemProps } />
```

File: packages/vuetify/src/components/VList/VListChildren.tsx [hunk 3]
```diff
@@ -24,25 +26,28 @@ export const VListChildren = genericComponent<new <T extends ListItem>() => {
...
+
+      if (type === 'subheader') return <VListSubheader {...itemProps} v-slots={ slots } />
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/VListChildren.tsx:36]:
```suggestion
      if (type === 'subheader') return <VListSubheader { ...itemProps } v-slots={ slots } />
```

File: packages/vuetify/src/components/VList/VListChildren.tsx [hunk 4]
```diff
@@ -24,25 +26,28 @@ export const VListChildren = genericComponent<new <T extends ListItem>() => {
...
-          value={value}
+          value={itemProps?.value}
           items={children}
           v-slots={{
             ...slots,
             header: headerProps => slots.externalHeader
-              ? slots.externalHeader({ ...rest, ...headerProps })
-              : <VListItem {...rest} {...headerProps} />,
+              ? slots.externalHeader({ ...itemProps, ...headerProps })
+              : <VListItem {...itemProps} {...headerProps} />,
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/VListChildren.tsx:46]:
```suggestion
              : <VListItem { ...itemProps } { ...headerProps } />,
```

File: packages/vuetify/src/components/VList/VListChildren.tsx [hunk 5]
```diff
@@ -24,25 +26,28 @@ export const VListChildren = genericComponent<new <T extends ListItem>() => {
...
           }}
         />
       ) : (
-        slots.item ? slots.item(item) : <VListItem {...item} v-slots={slots} />
+        slots.item ? slots.item(itemProps) : <VListItem {...itemProps} v-slots={ slots } />
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/VListChildren.tsx:50]:
```suggestion
        slots.item ? slots.item(itemProps) : <VListItem { ...itemProps } v-slots={ slots } />
```

File: packages/vuetify/src/components/VList/VListSubheader.tsx [hunk 6]
```diff
@@ -12,6 +12,8 @@ export const VListSubheader = defineComponent({
   props: {
     color: String,
     inset: Boolean,
+    title: String,
+    sticky: Boolean,
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/VListSubheader.tsx:16]:
```suggestion
    sticky: Boolean,
    title: String,
```

File: packages/vuetify/src/components/VList/list.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import { computed, inject, provide, ref } from 'vue'
```
Comment (johnleider) [inline packages/vuetify/src/components/VList/list.ts:1]:
```suggestion
// Utilities
import { computed, inject, provide, ref } from 'vue'
```

**Review by johnleider — CHANGES_REQUESTED** [review]
Looks good. I've made some suggestions regarding formatting, let me know what you think.

**Review by johnleider — CHANGES_REQUESTED** [review]
I think **list** could simply be a global composable, or even **depth**. Could use it to control the stacking order for various floating elements. Everything else looks good.

File: packages/vuetify/src/components/VList/VListSubheader.tsx [hunk 8]
```diff
@@ -13,30 +13,36 @@ export const VListSubheader = defineComponent({
     color: String,
     inset: Boolean,
     sticky: Boolean,
-    title: String,
+    text: String,
```
Comment (nekosaur) [inline packages/vuetify/src/components/VList/VListSubheader.tsx:16]: Why not title to keep it consistent with v-list-item?

Comment (johnleider) [reply packages/vuetify/src/components/VList/VListSubheader.tsx:16]: Because it's a subheading. By definition it wouldn't be a title.

Comment (nekosaur) [reply packages/vuetify/src/components/VList/VListSubheader.tsx:16]: > I think **list** could simply be a global composable, or even **depth**. Could use it to control the stacking order for various floating elements. Everything else looks good.

No need to make those changes now. Let's see if the need arises in the future

**Comment by nekosaur** [comment]
> I think **list** could simply be a global composable, or even **depth**. Could use it to control the stacking order for various floating elements. Everything else looks good.

No need to make those changes now, let's see if there's a need in the future

File: packages/vuetify/src/components/VList/VListSubheader.tsx (hunk 8 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VList/VListSubheader.tsx:16]: It's the title of the subheading? 🤷

**Review by johnleider — APPROVED** [review]

**Pull request merged**